### DISCUSSION TITLE
UrlCannotBeDetermined when storing files with custom file system disk

### DISCUSSION THREAD

**Discussion by LarryBarker** [spatie/laravel-medialibrary#2595] [Q&A]
Hello, first thank you for sharing this package. It is tremendously helpful. We had originally installed the latest version, only to find out our production environment would not support it so we had to downgrade to v7.x 

Since then, we have had some issues getting the package to work. Specifically, we are seeing the `UrlCannotBeDetermined` exception when trying to use `getFirstMediaUrl`. 

Here is our setup:

```
config/medialibrary.php

'disk_name' => env('MEDIA_DISK', 'public'),
```

```
.env

MEDIA_DISK=public
```

```
config/filesystems.php

'public' => [
            'driver' => 'local',
            'root' => storage_path('app/public'),
            'visibility' => 'public',
        ],
```

```
CreateResource.php

        // Add media from the request to the resource
        $resource->addMediaFromRequest('file')->toMediaCollection('resources');

        // Get the full URL for the media
        $link = $resource->getFirstMediaUrl('resources');
```

I have looked around for others with similar issues, and the closest I have found is: https://github.com/spatie/laravel-medialibrary/issues/86

so I added the debug statements to `LocalUrlGenerator`:

![image](https://user-images.githubusercontent.com/28734844/133905150-f115fb86-d644-4ba1-9335-3458fadc98d1.png)

So far, the only way I have been able to store files is to change the root for the disk from `storage_path('app/public')` to simply `public_path()`. The problem with this is that the files are being stored in the public directory, instead of being stored in the storage directory and symbolically linked to the public directory. 

Surely there is something I can do to fix this? I'd prefer to not have to write our own URL generator if possible. Any feedback or input is greatly appreciated :) Thank you!

**Comment by LarryBarker** [comment]
Just wanted to follow up here... if I change L65 in `LocalUrlGenerator` from:

```
        if (! Str::startsWith($this->getStoragePath(), public_path())) {
```

to:

```
        if (! Str::startsWith($this->getStoragePath(), storage_path())) {
```

Things work as expected. So it seems this version of the package requires the `public_path()` be used as a prefix for storing files?

**Reply by freekmurze** [reply]
I don't fully understand the problem, but I do welcome PRs for bug fixes. Make sure to add a test for it too 👍

**Comment by LarryBarker** [comment]
Hi Freek, thanks for your reply. The problem seems to be that
storage_path() helper cannot be used for the root file system because the
LocalUrlGenerator is expecting public_path() to be used.

I wouldn’t call it an error or bug if this is the expected behavior, just
trying to understand how to allow storage directory to be used.
